src/riscv/lib/src/pvm/linux.rs

```diff
@@ -1009,6 +1013,101 @@ impl<M: ManagerBase> SupervisorState<M> {
         // Return 0 as an indicator of success
         Ok(0)
     }
+
+    fn write_rlimit<MC: MemoryConfig>(
+        core: &mut MachineCoreState<impl MemoryConfig, M>,
+        addr: VirtAddr,
+        resource: parameters::Rlimit,
+    ) -> Result<(), Error>
+    where
+        M: ManagerReadWrite,
+    {
+        struct MemorySize<MC>(MC);
+        impl<MC: MemoryConfig> MemorySize<MC> {
+            const SIZE: u64 = MC::TOTAL_BYTES as u64;
+
+            const LOWER_BOUND: () = assert!(MC::TOTAL_BYTES >= u64::MIN as usize);
+            const UPPER_BOUND: () = assert!(MC::TOTAL_BYTES <= u64::MAX as usize);
+
+            // Hard limit on the size of the data segment
+            const RLIMIT_DATA: u64 = Self::SIZE;
+
+            // Hard limit on the size of the process stack
+            const RLIMIT_STACK: u64 = Self::SIZE;
+
+            // Hard limit on resident set size. Not used.
+            const RLIMIT_RSS: u64 = Self::SIZE;
+
+            // Hard limit on the memory that may be locked into RAM (B)
+            const RLIMIT_MEMLOCK: u64 = Self::SIZE;
+
+            // Hard limit on the size of a process's virtual memory (B)
+            const RLIMIT_AS: u64 = Self::SIZE;
+        }
+
+        let () = MemorySize::<MC>::LOWER_BOUND;
+        let () = MemorySize::<MC>::UPPER_BOUND;
+
+        let limit = match resource {
+            parameters::Rlimit::Cpu => parameters::RLIMIT_CPU,
+            parameters::Rlimit::Fsize => parameters::RLIMIT_FSIZE,
+            parameters::Rlimit::Data => MemorySize::<MC>::RLIMIT_DATA,
+            parameters::Rlimit::Stack => MemorySize::<MC>::RLIMIT_STACK,
+            parameters::Rlimit::Core => parameters::RLIMIT_CORE,
+            parameters::Rlimit::Rss => MemorySize::<MC>::RLIMIT_RSS,
+            parameters::Rlimit::Nproc => parameters::RLIMIT_NPROC,
+            parameters::Rlimit::Nofile => parameters::RLIMIT_NOFILE,
+            parameters::Rlimit::Memlock => MemorySize::<MC>::RLIMIT_MEMLOCK,
+            parameters::Rlimit::As => MemorySize::<MC>::RLIMIT_AS,
+            parameters::Rlimit::Locks => parameters::RLIMIT_LOCKS,
+            parameters::Rlimit::Sigpending => parameters::RLIMIT_SIGPENDING,
+            parameters::Rlimit::Msgqueue => parameters::RLIMIT_MSGQUEUE,
+            parameters::Rlimit::Nice => parameters::RLIMIT_NICE,
+            parameters::Rlimit::Rtprio => parameters::RLIMIT_RTPRIO,
+            parameters::Rlimit::Rttime => parameters::RLIMIT_RTTIME,
+        };
+
+        core.main_memory.write(addr.to_machine_address(), limit)?;
+        core.main_memory.write(
+            (addr + size_of_val(&limit) as u64).to_machine_address(),
```

Comment:
Do you mind using 8 instead? Or a constant for the size of `rlim_t` like some of the other system call handlers. I would like to avoid using host constants in PVM semantics.

```suggestion
            (addr + 8).to_machine_address(),
```